Add type hints to Individual class and fix detected issues

mschmidt87 opened the pull request:
This PR adds type hints to the `Individual` and `IndividualMultiGenome` classes and fixes the detected issues:

1. `Individual` and `IndividualMultiGenome` class
The relation between the two classes was not good: Having IndividualMultiGenome inherit from Individual causes problems. For instance, the self.genome member is a list of Genome instances in the former, but since it is initialized in the latter as a Genome instance, this does not make much sense. We could refrain from calling the super().__init__() function but then the inheritance does not make any sense. 

*Solution*: As discussed below, I removed the `IndividualMultiGenome` class and make the `self.genome` member of `Individual` a list. This triggered some changes in the `Population` class and the local search algorithms which now needs a list of `genome_params` and in the tests. As a side remark: Some functionality would not have worked for the multi genome version, e.g. the local search. This PR removes this problem.

2. The `Individual` class could be initialized with `genome=None` which is dangerous because all its methods rely on a genome being defined. I removed this option therefore and adapted the code in the `Population` class which had used this feature. 

3. `Individual.fitness` can be `None`
This is dangerous because some of the operations on the fitness relied on the fitness being a float without explicitly checking or catching exceptions. I remedied this in this PR.

mschmidt87 commented:
As a result, I detected the following issues with the current code:

- Having `IndividualMultiGenome` inherit from `Individual` causes problems. For instance, the `self.genome` member is a list of `Genome` instances in the former, but since it is initialized in the latter as a `Genome` instance, this does not make much sense. We could refrain from calling the `super().__init__()` function but then the inheritance does not make any sense. There are two possible solutions:
-- Make both independent classes and implement the common functions as external functions which are used by both classes. I implemented this for now.
-- Write an `IndividualBase` class and let the other two classes inherit from it. This would probably the cleanest solution

- The `Individual` class can currently be initialized with `genome=None` (happens in the `Population` class, see the diff), but this does not make sense, because the methods of `Individual` are unusable. That is why whenever it is initialized with `genome=None`, the randomize method is immediately called. `mypy` complains about this pattern which makes sense because this is ugly and dangerous. I would therefore propose the implemented solution. However, then we need to make a decision which rng we pass to the `randomize_genome` method if it is called in the `__init__` of `Individual`. All of this also extends to `IndividualMultiGenome`.

- The fitness value of individuals is now initialized as `None` and only set once it is computed. While this makes sense in principle, it causes many problems because some parts of the code rely on the fitness being a float (or an int), e.g. when we order individuals by their fitness.

FYI, this is the current outcome of `mypy`:
```
cgp/individual.py:109: error: Too few arguments for "randomize_genome" of "Individual"
cgp/population.py:130: error: Unsupported operand type for unary - ("Optional[float]")
cgp/ea/mu_plus_lambda.py:164: error: Unsupported operand type for unary - ("Optional[float]")
cgp/hl_api.py:79: error: Unsupported operand types for + ("None" and "float")
cgp/hl_api.py:79: note: Left operand is of type "Optional[float]"
```
The first line is related to the 2nd issue. The other ones are related to the third issue.
I'd like your feedback on all 3 issues @jakobj .

Also note that many tests fail at the moment, I will adapt them once we settled the issues above.

jakobj commented:
thanks for adding this! regarding your points:
1. what about replacing the `Individual` class with the `IndividualMultiGenome` class + some specific handling of the case where the `genome` argument is a dict, not a list of dicts? this seems to me the most natural solution as `Individual` is just a specific form of `IndividualMultiGenome`. otherwise i think i'm leaning towards the idea of having an `AbstractIndividual` base class that implements the shared functionality. both concrete classes then call these functions as required.
2. what about generating a `Genome` in the population class, then passing this as an argument to `Individual.__init__`? this seems natural to me as I see `Individual` just as a shallow wrapper around `Genome` anyway -- something that we could make even more explicit if you agree with this view. we would hence avoid the need to support `genome=None` in `Individual.__init__`
3. we use `fitness=None` as a signal to trigger evaluation of the individual on the objective, for example after non-silent mutation the fitness is set to `None`. I'm not sure how to implement this differently, but I'm open for ideas!

who knew such issues are lurking beneath such an innocent-sounding PR ;) mypy FTW!!!11

mschmidt87 commented:
> thanks for adding this! regarding your points:
> 
>     1. what about replacing the `Individual` class with the `IndividualMultiGenome` class + some specific handling of the case where the `genome` argument is a dict, not a list of dicts? this seems to me the most natural solution as `Individual` is just a specific form of `IndividualMultiGenome`. otherwise i think i'm leaning towards the idea of having an `AbstractIndividual` base class that implements the shared functionality. both concrete classes then call these functions as required.
> 
If we go for this, then I would not allow a single genome. Rather, an individual with a single genome has a list of genomes with only one entry. Then we could even call that class `Individual`. That means, in the current code we just make `Individual.genome` a list. 

>     2. what about generating a `Genome` in the population class, then passing this as an argument to `Individual.__init__`? this seems natural to me as I see `Individual` just as a shallow wrapper around `Genome` anyway -- something that we could make even more explicit if you agree with this view. we would hence avoid the need to support `genome=None` in `Individual.__init__`
> 
That would probably work, yes. I'll implement that.

>     3. we use `fitness=None` as a signal to trigger evaluation of the individual on the objective, for example after non-silent mutation the fitness is set to `None`. I'm not sure how to implement this differently, but I'm open for ideas!
> 

Yes, the `None` does make sense. But we need to handle the case that the fitness is `None` everywhere the fitness is actually used. `mypy` detected the four cases above. That will make the code a bit more verbose and lengthy. For instance, we could then check manually whether the fitness is `None` and throw an error if it is. What do you think?

jakobj commented:
> If we go for this, then I would not allow a single genome. Rather, an individual with a single genome has a list of genomes with only one entry. Then we could even call that class `Individual`. That means, in the current code we just make `Individual.genome` a list.

sounds good to me!

> Yes, the `None` does make sense. But we need to handle the case that the fitness is `None` everywhere the fitness is actually used. `mypy` detected the four cases above. That will make the code a bit more verbose and lengthy. For instance, we could then check manually whether the fitness is `None` and throw an error if it is. What do you think?

I think it's a good idea to add these checks as they provide an additional layer of safety against strange things happening without an error being raised, even if that means more loc

mschmidt87 commented:
I addressed all the issues and documented the solutions in the PR description above. Please take another look @jakobj

jakobj commented on cgp/individual.py:
```diff
@@ -46,19 +89,7 @@ def crossover(self, other_parent, rng):
-
-    def mutate(self, mutation_rate, rng):
+    def mutate(self, mutation_rate: float, rng: np.random.RandomState) -> None:
         """Mutate the individual in place.
```
```suggestion
        """Mutate the individual.
```

mschmidt87 replied:
In place is correct, isn't it? The function does create a new individual.

jakobj replied:
yes, technically it's correct, but isn't it a bit redundant to state for every method that/whether it operates on the corresponding class? in contrast, if this would be a regular function that receives an individual, this additional info would be very helpful!

jakobj commented on cgp/individual.py:
```diff
@@ -90,89 +124,51 @@ def randomize_genome(self, genome_params, rng):
+        Callable[[List[float]], List[float]]
         """
-        return CartesianGraph(self.genome).to_func()
+        return [CartesianGraph(g).to_func() for g in self.genome]
```
i feel like this will add a lot of unpacking for the user, won't it? it's already suboptimal that they will need to do `f([...])[0]` for a function with a single output, now it would read `f[0]([...])[0]`. this is not very pretty. should we add automatic unpacking for the special case of a single genome, even if this seems to contradict best practice? this holds essentially for all cases where we add an additional list in this new implementation

jakobj commented on cgp/local_search/gradient_based.py:
```diff
@@ -57,20 +57,25 @@ def gradient_based(
+    for torch_mod in f:
+        params += list(torch_mod.parameters())
+    if len(params) > 0:
+        optimizer = optimizer_class(params, lr=lr)  #
```
```suggestion
        optimizer = optimizer_class(params, lr=lr)
```

jakobj commented on cgp/population.py:
```diff
@@ -1,8 +1,8 @@
-
-from .individual import Individual, IndividualMultiGenome
+from typing import List, Union
+from .genome import Genome
```
```suggestion

from .genome import Genome
```

jakobj commented on cgp/population.py:
```diff
@@ -77,17 +77,11 @@ def _label_new_individuals(self, individuals: List[Individual]) -> None:
-            else:
-                raise NotImplementedError()
-            individual.randomize_genome(self._genome_params, self.rng)
+            genome = [Genome(**gd) for gd in self._genome_params]
```
maybe use `genomes` as variable name whereever this is a list?

jakobj commented on cgp/population.py:
```diff
@@ -123,7 +117,14 @@ def crossover(self, breeding_pool: List[Individual], n_offsprings: int) -> List[
+
+        def sort_func(ind: Individual) -> float:
+            if isinstance(ind.fitness, float):
+                return -ind.fitness
```
same as above, use `reversed` instead

mschmidt87 replied:
Done. And added a comment.

jakobj commented on cgp/individual.py:
```diff
@@ -1,43 +1,84 @@
+    """Base class for all individuals.
+    """
+
+    def __init__(self, fitness: Union[float, None], genome: Union[Genome, List[Genome]]) -> None:
```
to avoid the unions for genome, wouldn't it be cleaner to only define this member in the derived classes?

mschmidt87 replied:
this is needed by the `clone` function.

jakobj commented on cgp/individual.py:
```diff
@@ -1,43 +1,84 @@
         Individual
         """
-        return Individual(self.fitness, self.genome.clone())
+        return type(self)(self.fitness, self.genome)
```
```suggestion
        raise NotImplementedError()
```

jakobj replied:
it does not make sense to use an instance of `IndividualBase`, does it? so i would just not provide any implementation for functions that are overridden in derived classes

mschmidt87 replied:
the idea is that we don't need to define the `clone` function in the child classes.

jakobj commented on cgp/individual.py:
```diff
@@ -46,35 +87,23 @@ def crossover(self, other_parent, rng):
         None
         """
-        self._mutate(self.genome, mutation_rate, rng)
+        pass
```
```suggestion
        raise NotImplementedError()
```

jakobj commented on cgp/ea/mu_plus_lambda.py:
```diff
@@ -159,20 +160,29 @@ def _sort(self, combined: List[Individual]) -> List[Individual]:
                 ind.fitness = -np.inf
 
-        # get list of indices that sorts combined_copy ("argsort")
+        # get list of indices that sorts combined_copy ("argsort") in descending order
```
i would move this comment below. mentally i tend to skip functions until i see where there are used and that's where i would need the information that line 173 does argsort

mschmidt87 replied:
Done

jakobj commented on cgp/local_search/gradient_based.py:
```diff
@@ -57,20 +57,33 @@ def gradient_based(
 
-    if len(list(f.parameters())) > 0:
-        optimizer = optimizer_class(f.parameters(), lr=lr)
+    if isinstance(f, List):
```
```suggestion
    if isinstance(f, list):
```

mschmidt87 replied:
Done.

jakobj commented on cgp/population.py:
```diff
@@ -48,61 +53,67 @@ def __init__(
+                genome.randomize(self.rng)
+                individual_s = IndividualSingleGenome(
+                    fitness=None, genome=genome
+                )  # type: IndividualBase
```
```suggestion
                )
```

jakobj replied:
or what is this needed for?

mschmidt87 replied:
It is needed by mypy. I added an explanatory comment behind this.

jakobj commented on cgp/population.py:
```diff
@@ -48,61 +53,67 @@ def __init__(
+                    gen.randomize(self.rng)
+                individual_m = IndividualMultiGenome(
+                    fitness=None, genomes=genomes
+                )  # type: IndividualBase
```
```suggestion
                )
```

mschmidt87 replied:
see above.

mschmidt87 commented:
Addressed all your comments, @jakobj . One more point: I would rename `IndividualMultiGenome.genomes` to `genome`. It might be orthographically more correct to say `genomes`, but it makes it more difficult to handle for the user if they want to switch from SingleGenome to MultiGenome. It also introduces more complex code to handle the different variable names in the library and in the tests.

jakobj commented on cgp/local_search/gradient_based.py:
```diff
@@ -57,20 +57,33 @@ def gradient_based(
 
-    if len(list(f.parameters())) > 0:
-        optimizer = optimizer_class(f.parameters(), lr=lr)
+    if isinstance(f, list):
```
this stuff makes me flinch every time i see it, but i don't think we have an easier solution. oh well :|

jakobj commented on cgp/population.py:
```diff
@@ -48,61 +53,71 @@ def __init__(
-            individuals.append(individual)
-
+                genomes: List[Genome] = [Genome(**gd) for gd in self._genome_params]
+                for gen in genomes:
+                    gen.randomize(self.rng)
```
```suggestion
                for g in genomes:
                    g.randomize(self.rng)
```

jakobj commented on cgp/population.py:
```diff
@@ -48,61 +53,71 @@ def __init__(
+                individual_m = IndividualMultiGenome(
+                    fitness=None, genomes=genomes
+                )  # type: IndividualBase # indicates to mypy that
+                # individual_s is an instance of a child class of
```
```suggestion
                # individual_m is an instance of a child class of
```

mschmidt87 commented on cgp/hl_api.py:
```diff
@@ -63,6 +63,7 @@ def evolve(
         # progress printing, recording, checking exit condition etc.; needs to
         # be done /after/ new parent population was populated from combined
         # population and /before/ new individuals are created as offsprings
+        assert isinstance(pop.champion.fitness, float)
```
Yes, necessary.

mschmidt87 replied:
necesary.

mschmidt87 commented on cgp/individual.py:
```diff
@@ -1,178 +1,223 @@
-        Callable
+    @staticmethod
+    def _update_parameters_from_torch_class(genome: Genome, torch_cls: "torch.nn.Module") -> bool:
+        """Update parameters of the individual from a torch class.
```
done.

mschmidt87 replied:
done.

mschmidt87 commented:
Addressed all comments. @jakobj

jakobj commented on test/test_hl_api.py:
```diff
@@ -168,7 +166,7 @@ def test_speedup_parallel_evolve(population_params, genome_params, ea_params):
 
     for n_processes in [1, 2, 4]:
-        pop = cgp.Population(**population_params, genome_params=genome_params)
+        pop = cgp.Population(**population_params, genome_params=[genome_params])
```
why does this need to be a list of params?

mschmidt87 replied:
It doesn't. I reverted the change.

jakobj commented:
excellent work :sparkles: :tada: :sparkles:  and merging
